### DISCUSSION TITLE
uploaded files do not display in pages in wikijs

### DISCUSSION THREAD

**Discussion by ghost** [requarks/wiki#3322] [Error / Bug Report]
Describe the bug
Can't access to uploaded files: path must be absolute or specify root to res.sendFile

**`0|server | 2021-02-06T12:50:35.499Z [MASTER] error: path must be absolute or specify root to res.sendFile`**

Equals to this problem: #2463 and #2930

To Reproduce
Administration -> Utilities -> Flush Pages and Assets Cache

Expected behavior
These videos and pictures do not display 

Screenshots
![image](https://user-images.githubusercontent.com/59251666/107141473-2552bc00-693e-11eb-88a1-27a6605f392b.png)

![Report](https://user-images.githubusercontent.com/59251666/107141788-5fbd5880-6940-11eb-9d20-2920524169ac.png)

Host Info (please complete the following information):
OS: Linux - Debian (buster) 10 x64
Wiki.js version: 2.5.159
Database engine: PostgreSQL 11.4

**Comment by ghost** [comment]
Suggested solution:


this line:
```yaml
await sendFile(cachePath, { dotfiles: 'deny', root: "/opt/wiki/" })
```


replace with this line:
```yaml
try {
await sendFile(cachePath, { dotfiles: 'deny' })
}catch (error) {
await sendFile(cachePath, { dotfiles: 'deny', root: '/opt/wiki/' })
}
```

**Reply by tmunsch** [reply]
@Fakhri-Hasanzadeh What file are you referring to ? I am having the same issue and desperately need to fix this, just spent 6 hours building this wiki, i restart the server and then this issue arises. Please help

**Reply by kanebrands** [reply]
This is in server/models/assets.js


I made this change on my system with the exception of the root path has to match your system:

`
  static async getAssetFromCache(assetPath, cachePath, res) {

    try {

      await fs.access(cachePath, fs.constants.R_OK)

    } catch (err) {

      return false

    }

    const sendFile = Promise.promisify(res.sendFile, {context: res})

    res.type(path.extname(assetPath))

    //await sendFile(cachePath, { dotfiles: 'deny' })

    try {

      await sendFile(cachePath, { dotfiles: 'deny' })

    }catch (error) {

      await sendFile(cachePath, { dotfiles: 'deny', root: '/var/www/wiki/' })

    }

    return true

  }

`

This is the function that required modification -- if I were a better programming I would make the config.yml file have an absolute path in it and just reference that variable here but I am not a JS programmer.

**Comment by nswa203** [comment]
I have the same symptoms.
Just started creating a site within a docker stack. 
Everything initially looked fine. However,  two days later all the locally uploaded assets are no longer visible.
Network trace shows: status 500 Internal Server Error for each of the assets. 
Wiki console shows: 2022-03-11T11:38:27.136Z [MASTER] error: path must be absolute or specify root to res.sendFile

I've taken a quick look at the sql db using phpmyadmin and the  asset entries appear to look fine.

So ... I deleted the docker container and started again. Reinstalled and rebuilt a few pages.  Exactly the same scenario, all looked OK, then after a couple of days all the assets disappeared again! Very odd indeed.
I have many other docker containers running on this server and all are working fine ... this one has got me stumped.

Can you help me diagnose this problem please? I'm happy to try any patch/test version as this is not a production service.

Current Version 2.5.276
I've tried a couple of databases - current is 10.5.13-MariaDB-log mysql

**Comment by nswa203** [comment]
What a waste of time ....

Turns out to be a permissions issue.
I always build my docker containers with my admin's PUID & PGID unless the product's docker implementation documentation says otherwise. Reverting to root's PUID=0 & PGID=0 has fixed the issue. 

I suspect that when you upload an asset it goes straight into cache and is subsequently loaded from the cache and all looks OK. Unfortunately as time goes by and the cache gets cleared, the asset has to be fetched from disk storage with all the problematic permissions checking etc...

I'm not really happy running a container with root permissions, but will do while I test further.

**Comment by rgrunbla** [comment]
Hello,

I had this problem on my wiki, as well as images being lost between restart (the cache wasn't being regenerated correctly).

The root cause was that the path in the "Local File System" configuration option from wiki-js was a relative path of the form "./foobar", and not an absolute one, as required.

**Reply by lightrush** [reply]
To clarify, this is referring to the Storage > Local File System config. If you don't use it you can turn it off, which also resolves the issue.

**Comment by jcdufourd** [comment]
I had the same problem of images not appearing, and solved it by changing the "Path" of the "Local File System" in the "Storage" section of the admin area from **_relative_** to **_absolute_** (as required in the comment below the field)
